File: csharp/src/refactoring/golden-master/GoldenMaster.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace RefactoringAvanzado.Refactoring.GoldenMaster;

public class OrderItem
{
    public string Sku { get; init; } = string.Empty;
    public string Description { get; init; } = string.Empty;
    public decimal UnitPrice { get; init; }
    public int Quantity { get; init; }
    public string? Category { get; init; }
}

public class Order
{
    public string Id { get; init; } = string.Empty;
    public string CustomerName { get; init; } = string.Empty;
    public List<OrderItem> Items { get; init; } = new();
}

public class ReceiptPrinter
{
    // Do not change this function at the beginning of the exercise; first create the Golden Master.
    public string Print(Order order)
    {
        var now = GetCurrentDate();

        var header = $"Recibo {order.Id} - {now:dd/MM/yyyy} {now:HH:mm:ss}";

        decimal subtotal = 0;
        var lines = order.Items.Select((it, idx) =>
        {
            var lineTotal = Round(it.UnitPrice * it.Quantity);
            subtotal = Round(subtotal + lineTotal);
            return $"{idx + 1}. {it.Description} ({it.Sku}) x{it.Quantity} = ${lineTotal:F2}";
        }).ToList();

        var luckyDiscountPct = Discount();
        var luckyDiscount = Round(subtotal * luckyDiscountPct);

        var taxableGeneral = order.Items
            .Where(i => i.Category != "books")
            .Sum(i => i.Category == "food" ? 0 : i.UnitPrice * i.Quantity);

        var foodTax = order.Items
            .Where(i => i.Category == "food")
            .Sum(i => i.UnitPrice * i.Quantity * 0.03m);

        var generalTax = taxableGeneral * 0.07m;
        var taxes = Round(generalTax + foodTax);

        var total = Round(subtotal - luckyDiscount + taxes);

        var summary = new List<string>
        {
            $"Subtotal: ${subtotal:F2}",
            luckyDiscount > 0
                ? $"Descuento de la suerte: -${luckyDiscount:F2} ({luckyDiscountPct * 100:F2}%)"
                : "Descuento de la suerte: $0.00 (0.00%)",
            $"Impuestos: ${taxes:F2}",
            $"TOTAL: ${total:F2}"
        };

        var allLines = new List<string> { header };
        allLines.AddRange(lines);
        allLines.Add("---");
        allLines.AddRange(summary);

        return string.Join("\n", allLines);
    }

    protected virtual decimal Discount()
    {
        var random = new Random();
        decimal luckyDiscountPct = 0;
        if (random.NextDouble() < 0.1)
        {
            luckyDiscountPct = (decimal)random.NextDouble() * 0.05m;
        }
        return luckyDiscountPct;
    }

    protected virtual DateTime GetCurrentDate()
    {
        return DateTime.Now;
    }

    private static decimal Round(decimal n)
    {
        return Math.Round(n, 2);
    }
}

public static class OrderGenerator
{
    private static readonly List<OrderItem> Products = new()
    {
        new OrderItem { Sku = "BK-001", Description = "Libro: Clean Code", UnitPrice = 30, Category = "books" },
        new OrderItem { Sku = "FD-010", Description = "Café en grano 1kg", UnitPrice = 12.5m, Category = "food" },
        new OrderItem { Sku = "GN-777", Description = "Cuaderno A5", UnitPrice = 5.2m, Category = "general" },
        new OrderItem { Sku = "GN-123", Description = "Bolígrafos (pack 10)", UnitPrice = 3.9m, Category = "general" },
        new OrderItem { Sku = "FD-222", Description = "Té verde 200g", UnitPrice = 6.75m, Category = "food" }
    };

    public static readonly List<string> Customers = new() { "Ana", "Luis", "Mar", "Iván", "Sofía" };

    // Utility to generate Orders
    public static Order GenerateOrder(string id, string customerName, int numItems, int quantity)
    {
        var items = new List<OrderItem>();
        for (int i = 0; i < numItems; i++)
        {
            var p = Products[i];
            items.Add(new OrderItem
            {
                Sku = p.Sku,
                Description = p.Description,
                UnitPrice = p.UnitPrice,
                Category = p.Category,
                Quantity = quantity
            });
        }

        return new Order { Id = id, CustomerName = customerName, Items = items };
    }
}

File: typescript/src/refactoring/golden-master/golden-master.ts
export type OrderItem = {
  sku: string
  description: string
  unitPrice: number
  quantity: number
  category?: 'general' | 'food' | 'books'
}

export type Order = {
  id: string
  customerName: string
  items: OrderItem[]
}

export class ReceiptPrinter {
  // Do not change this function at the beginning of the exercise; first create the Golden Master.
  print(order: Order): string {
    const now = this.getCurrentDate()

    const header = `Recibo ${order.id} - ${now.toLocaleDateString()} ${now.toLocaleTimeString()}`

    let subtotal = 0
    let lines = order.items.map((it, idx) => {
      const lineTotal = round(it.unitPrice * it.quantity)
      subtotal = round(subtotal + lineTotal)
      return `${idx + 1}. ${it.description} (${it.sku}) x${it.quantity} = $${lineTotal.toFixed(2)}`
    })
    let luckyDiscountPct = this.discount();
    const luckyDiscount = round(subtotal * luckyDiscountPct)

    const taxableGeneral = order.items
      .filter((i) => i.category !== 'books')
      .reduce((s, i) => s + (i.category === 'food' ? 0 : i.unitPrice * i.quantity), 0)
    const foodTax = order.items
      .filter((i) => i.category === 'food')
      .reduce((s, i) => s + i.unitPrice * i.quantity * 0.03, 0)
    const generalTax = taxableGeneral * 0.07
    const taxes = round(generalTax + foodTax)

    const total = round(subtotal - luckyDiscount + taxes)

    const summary = [
      `Subtotal: $${subtotal.toFixed(2)}`,
      luckyDiscount > 0
        ? `Descuento de la suerte: -$${luckyDiscount.toFixed(2)} (${(luckyDiscountPct * 100).toFixed(2)}%)`
        : `Descuento de la suerte: $0.00 (0.00%)`,
      `Impuestos: $${taxes.toFixed(2)}`,
      `TOTAL: $${total.toFixed(2)}`,
    ]

    return [header, ...lines, '---', ...summary].join('\n')
  }

  protected discount() {
    let luckyDiscountPct = 0
    if (Math.random() < 0.1) {
      luckyDiscountPct = Math.random() * 0.05
    }
    return luckyDiscountPct;
  }

  protected getCurrentDate() {
    return new Date(Date.now());
  }
}

function round(n: number): number {
  return Math.round(n * 100) / 100
}

const products: Omit<OrderItem, 'quantity'>[] = [
  {sku: 'BK-001', description: 'Libro: Clean Code', unitPrice: 30, category: 'books'},
  {sku: 'FD-010', description: 'Café en grano 1kg', unitPrice: 12.5, category: 'food'},
  {sku: 'GN-777', description: 'Cuaderno A5', unitPrice: 5.2, category: 'general'},
  {sku: 'GN-123', description: 'Bolígrafos (pack 10)', unitPrice: 3.9, category: 'general'},
  {sku: 'FD-222', description: 'Té verde 200g', unitPrice: 6.75, category: 'food'},
]

const customers = ['Ana', 'Luis', 'Mar', 'Iván', 'Sofía']

// Utility to generate Orders
export function generateOrder(
  id: string,
  customerName: string,
  numItems: number,
  quantity: number,
): Order {
  const items: OrderItem[] = []
  for (let i = 0; i < numItems; i++) {
    const p = products[i]
    items.push({...p, quantity: quantity} as OrderItem) // 1..4 unidades
  }

  return {id, customerName, items}
}
